File: tests/unit/test_timer.py
import cocotb
from cocotb.triggers import Timer, RisingEdge
from cocotb.clock import Clock
import sys
import os

# Import infrastructure
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from infrastructure import run_test_simple

@cocotb.test()
async def timer_test(dut):
    """Test Timer Peripheral."""

    # Start Clock
    cocotb.start_soon(Clock(dut.clk, 10, unit="ns").start())

    # Reset
    dut.rst_n.value = 0
    dut.write_enable.value = 0
    dut.address.value = 0
    dut.write_data.value = 0
    
    await Timer(20, unit="ns")
    dut.rst_n.value = 1
    await Timer(10, unit="ns")

    # Addresses
    MTIME_L    = 0x40004000
    MTIME_H    = 0x40004004
    MTIMECMP_L = 0x40004008
    MTIMECMP_H = 0x4000400C

    # 1. Check Initial State
    # mtime should be small (reset to 0 and counting)
    dut.address.value = MTIME_L
    await Timer(1, unit="ns")
    mtime_l = int(dut.read_data.value)
    assert mtime_l < 100, "mtime should start near 0"
    
    # mtimecmp should be max
    dut.address.value = MTIMECMP_L
    await Timer(1, unit="ns")
    assert int(dut.read_data.value) == 0xFFFFFFFF, "mtimecmp_l init failed"
    
    dut.address.value = MTIMECMP_H
    await Timer(1, unit="ns")
    assert int(dut.read_data.value) == 0xFFFFFFFF, "mtimecmp_h init failed"
    
    # Interrupt should be 0
    assert dut.interrupt_request.value == 0, "Interrupt should be 0 initially"

    # 2. Set Compare Value
    # Set mtimecmp to current mtime + 20
    # Read current mtime again to be sure
    dut.address.value = MTIME_L
    await Timer(1, unit="ns")
    current_time = int(dut.read_data.value)
    
    target_time = current_time + 40 # Wait 40 cycles
    
    # Write MTIMECMP_L
    dut.address.value = MTIMECMP_L
    dut.write_data.value = target_time
    dut.write_enable.value = 1
    await RisingEdge(dut.clk)
    
    # Write MTIMECMP_H (0)
    dut.address.value = MTIMECMP_H
    dut.write_data.value = 0
    await RisingEdge(dut.clk)
    dut.write_enable.value = 0
    
    # 3. Wait for Interrupt
    # Wait enough cycles (target_time - current_time + margin)
    for _ in range(100):
        await RisingEdge(dut.clk)
        if dut.interrupt_request.value == 1:
            break
        
    assert dut.interrupt_request.value == 1, f"Interrupt did not fire. Target={target_time}"

    # 4. Clear Interrupt (by increasing mtimecmp)
    dut.address.value = MTIMECMP_L
    dut.write_data.value = 0xFFFFFFFF
    dut.write_enable.value = 1
    await RisingEdge(dut.clk)
    dut.write_enable.value = 0
    await Timer(1, unit="ns")
    
    assert dut.interrupt_request.value == 0, "Interrupt did not clear"

    dut._log.info("Timer Test Passed!")

def test_timer():
    run_test_simple(
        module_name="test_timer",
        toplevel="timer",
        rtl_files=["peripherals/timer.v"],
        file_path=__file__
    )

File: test/hardware/unit/test_timer.cpp
#include "tb_base.h"
#include "Vtimer.h"
#include <string>

class TimerTestbench : public ClockedTestbench<Vtimer> {
public:
    static constexpr uint32_t MTIME_L    = 0x40004000;
    static constexpr uint32_t MTIME_H    = 0x40004004;
    static constexpr uint32_t MTIMECMP_L = 0x40004008;
    static constexpr uint32_t MTIMECMP_H = 0x4000400C;
    
    TimerTestbench() : ClockedTestbench<Vtimer>(100, true, "timer_trace.vcd") {
        // Initialize inputs
        dut->write_enable = 0;
        dut->address = 0;
        dut->write_data = 0;
    }
    
    void set_clk(uint8_t value) override {
        dut->clk = value;
    }
    
    void reset() {
        dut->rst_n = 0;
        tick();
        dut->rst_n = 1;
        tick();
        TB_LOG("Reset complete");
    }
    
    uint32_t read_reg(uint32_t addr) {
        dut->address = addr;
        eval();
        return dut->read_data;
    }
    
    void write_reg(uint32_t addr, uint32_t data) {
        dut->address = addr;
        dut->write_data = data;
        dut->write_enable = 1;
        tick();
        dut->write_enable = 0;
    }
    
    void test_initial_state() {
        TB_LOG("Test: Initial state");
        
        // mtime should be small (recently reset)
        uint32_t mtime_l = read_reg(MTIME_L);
        if (mtime_l >= 100) {
            std::string msg = "mtime should start near 0, got " + std::to_string(mtime_l);
            throw std::runtime_error(msg);
        }
        
        // mtimecmp should be max
        TB_ASSERT_EQ(read_reg(MTIMECMP_L), 0xFFFFFFFF, "mtimecmp_l init");
        TB_ASSERT_EQ(read_reg(MTIMECMP_H), 0xFFFFFFFF, "mtimecmp_h init");
        
        // Interrupt should be low
        TB_ASSERT_EQ(dut->interrupt_request, 0, "Initial interrupt");
    }
    
    void test_interrupt_trigger() {
        TB_LOG("Test: Interrupt trigger");
        
        // Read current time
        uint32_t current_time = read_reg(MTIME_L);
        uint32_t target_time = current_time + 40;
        
        // Set mtimecmp
        write_reg(MTIMECMP_L, target_time);
        write_reg(MTIMECMP_H, 0);
        
        // Wait for interrupt to fire
        bool interrupt_fired = false;
        for (int i = 0; i < 100; i++) {
            tick();
            if (dut->interrupt_request == 1) {
                interrupt_fired = true;
                TB_LOG("Interrupt fired after waiting");
                break;
            }
        }
        
        if (!interrupt_fired) {
            std::string msg = "Interrupt did not fire. Target=" + std::to_string(target_time);
            throw std::runtime_error(msg);
        }
    }
    
    void test_interrupt_clear() {
        TB_LOG("Test: Clear interrupt");
        
        // Interrupt should still be asserted from previous test
        TB_ASSERT_EQ(dut->interrupt_request, 1, "Interrupt before clear");
        
        // Clear by setting mtimecmp to max
        write_reg(MTIMECMP_L, 0xFFFFFFFF);
        eval();
        
        TB_ASSERT_EQ(dut->interrupt_request, 0, "Interrupt after clear");
    }
};

int main(int argc, char** argv) {
    Verilated::commandArgs(argc, argv);
    
    try {
        TimerTestbench tb;
        
        tb.reset();
        tb.test_initial_state();
        tb.test_interrupt_trigger();
        tb.test_interrupt_clear();
        
        TB_LOG("All Timer tests PASSED!");
        return 0;
        
    } catch (const std::exception& e) {
        TB_ERROR(e.what());
        return 1;
    }
}
